### DISCUSSION TITLE
Multimonitor setup, renaming workspaces etc

### DISCUSSION THREAD

**Discussion by xoores** [hyprwm/Hyprland#3364] [Q&A]
Hi,

I'm "migrating" from i3/Sway to Hyprland and in the process I ran into few issues here and there - probably they are simple to solve but I was not able to figure out how. So I'm trying here :slightly_smiling_face: 

## 1. Multimonitor setup
I have 3 monitors in a setup like this: `[Screen1][Screen2][LaptopScreen]`. That means that I wanna have my laptop on the very right of the setup. I think that I cannot do that purely from config (as I would have to configure constant offset to laptop screen even when no external monitors are connected).

I figured out that I would probably do "on-the-fly" configuration using `hyprctl keyword monitor=xxxxxxxx` to configure everything.

Is this approach correct or am I missing something?

## 2. Cannot apply rules on renamed workspace
I have a script that renames workspaces - I use it to create "icons" of applications on specific workspaces like this:
![image](https://github.com/hyprwm/Hyprland/assets/2182467/ba52eccb-f76a-465d-b729-bcdb5691f064)

However when I rename workspace (the "globe" one is ID=1), Hyprland stops applying rules to that workspace. I have this in my config:
`workspace = 1,gapsin:0,gapsout:0,rounding:false,bordersize:0` but it is ignored - even though the workspace 1 exists:
```
workspace ID 1 () on monitor eDP-1:
	windows: 1
	hasfullscreen: 0
	lastwindow: 0x558db865b4c0
	lastwindowtitle: New Discussion · hyprwm/Hyprland - Vivaldi
```

Am I doing something wrong or is this a bug? :thinking: 

## 3. Some apps are launched with delay
Some apps (like kitty) have noticeable delay before they open their window. I've read the FAQ etc and there is a mention that this is caused by the desktop-portal not running, but I already have these launched so I would think that this is not the problem.

Part of my script that is executed by `exec-once`:
```
killall xdg-desktop-portal-hyprland
killall xdg-desktop-portal-gnome
killall xdg-desktop-portal-wlr
killall xdg-desktop-portal-gtk
killall xdg-desktop-portal

dbus-update-activation-environment WAYLAND_DISPLAY XDG_CURRENT_DESKTOP 

killall waybar
(
	launch /usr/libexec/xdg-desktop-portal-hyprland
	sleep 2
	launch /usr/libexec/xdg-desktop-portal
) &
```
Processes:
```
xoores    1899  0.0  0.0 605800  5800 tty1     Sl+  Sep17   0:00 /usr/libexec/xdg-document-portal
xoores    1907  0.0  0.0 308316  5792 tty1     Sl+  Sep17   0:00 /usr/libexec/xdg-permission-store
xoores    9697  0.0  0.0   2512  1644 tty1     S+   Sep17   0:00 /usr/bin/bwrap --args 66 -- /usr/bin/xdg-dbus-proxy --args=63
xoores    9698  0.0  0.0   2512  1408 tty1     S+   Sep17   0:00 /usr/bin/bwrap --args 66 -- /usr/bin/xdg-dbus-proxy --args=63
xoores    9699  0.0  0.0 234464  4848 tty1     Sl+  Sep17   0:00 /usr/bin/xdg-dbus-proxy --args=63
xoores   12457  0.0  0.0 531192  7020 tty1     Sl+  Sep17   0:00 /usr/libexec/xdg-desktop-portal
xoores   12464  0.0  0.0 411276 24224 tty1     Sl+  Sep17   0:01 /usr/libexec/xdg-desktop-portal-gtk
xoores   12495  0.0  0.0 945496 54868 tty1     Sl+  Sep17   0:00 /usr/libexec/xdg-desktop-portal-hyprland
```

What am I missing or what am I doing wrong? :thinking: 

## 4. Move workspace to monitor (via keybinding)
In i3 I have a keybinding that moves active workspace to the "left" or "right":
```
bindsym $mod+Control+Left move workspace to output left
bindsym $mod+Control+Right move workspace to output right
```

Meaning not to the specific monitor but to the relative direction. How can I achieve the same behavior in Hyprland? I know that there is a `movecurrentworkspacetomonitor` but I could not figure out how to do "relative" moving - I really *don't* want/need to create a keybinding for each connected monitor :sweat_smile: I did not see anything relevant in the documentation nor on Google...

## 5. Waybar ignores changes of workspace names
I'm currently unclear if this belongs here or to Waybar, but when I rename a workspace, I would somehow expect Waybar to reflect that change. It doesn't.

Waybar config (relevant part) is:
```
"hyprland/workspaces": {
    "all-outputs": false,
    "active-only": false,
    "on-click": "activate",
    "sort-by-name": false,
    "sort-by-number": true,
    "on-scroll-up": "hyprctl dispatch workspace e+1",
    "on-scroll-down": "hyprctl dispatch workspace e-1",
    "format": "{name}",
},
```

Currently I'm sending SIGUSR2 to Waybar every time I rename workspace. SIGUSR2 forces Waybar to reload config - including the workspaces from Hyprland. The downside of this is that Waybar "blinks" - it disappears & reappears, which is quite ugly and distracting. I would obvously like to achieve this in a more elegant way :grin: 

Did anyone tried to solve something similar? Is this a problem in Hyprland not triggering some kind of event towards Waybar or is this a Waybar bug/problem? If this is a Waybar problem, I will open ticket there of course - just wanted to ask, just in case...

**Comment by pbosab** [comment]
> ## 1. Multimonitor setup

https://wiki.hyprland.org/Configuring/Monitors/#general

using this as example: 

```
monitor=name,resolution,position,scale
```

for your config (change with the correct values):
```
monitor=[Screen1Port],[Screen1ResX]x[Screen1ResY]@[Screen1RefreshRate],0x0,1
monitor=[Screen2Port],[Screen2ResX]x[Screen2ResY]@[Screen2RefreshRate],[Screen1ResX]x0,1
monitor=[Screen3Port],[Screen3ResX]x[Screen3ResY]@[Screen3RefreshRate],[Screen1ResX+Screen2ResX]x0,1
```

> ## 2. Workspace renaming

[hyprland-autoname-workspace](https://github.com/hyprland-community/hyprland-autoname-workspaces)

> ## 3. Some apps are launched with delay

:thinking: 

> ## 4. Move workspace to monitor (via keybinding)

maybe creating 3 binds with swapactiveworkspaces? something like this:

```
bind = $mainMod TAB, 1, swapactiveworkspaces, 0 1
bind = $mainMod TAB, 2, swapactiveworkspaces, 1 2
bind = $mainMod TAB, 3, swapactiveworkspaces, 0 2
```

> ## 5. Waybar ignores changes of workspace names

don't know at which point are their releases, but using hyprland-git and waybar-git this works.
there were changes about waybar modules and hyprland IPC lately...

**Reply by xoores** [reply]
Thanks for the reply :slightly_smiling_face: 

## 1. Multimonitor setup
> 
> https://wiki.hyprland.org/Configuring/Monitors/#general
> 
> using this as example:
> 
> ```
> monitor=name,resolution,position,scale
> ```
> 
> for your config (change with the correct values):
> 
> ```
> monitor=[Screen1Port],[Screen1ResX]x[Screen1ResY]@[Screen1RefreshRate],0x0,1
> monitor=[Screen2Port],[Screen2ResX]x[Screen2ResY]@[Screen2RefreshRate],[Screen1ResX]x0,1
> monitor=[Screen3Port],[Screen3ResX]x[Screen3ResY]@[Screen3RefreshRate],[Screen1ResX+Screen2ResX]x0,1
> ```
I know, I did that.

The thing is that my laptop is then configured with that "offset" of 6400 (as `monitor=eDP1,3840x2160@120,6400x0,1`) even if there are no other monitors attached.

And if I attach just a single monitor, it does not really make sense either - that's why I was wondering whether "dynamic" way is the correct approach here.


## 2. Workspace renaming / Cannot apply rules on renamed workspace
> 
> [hyprland-autoname-workspace](https://github.com/hyprland-community/hyprland-autoname-workspaces)

I already have something that does the renaming, but thanks for this, I'll take a look!

The problem is that *after renaming* Hyprland stops applying rules from the config to that workspace :slightly_smiling_face: If I write `workspace = 1,....` it gets ignored, if I try `workspace = name:xxxxx` it does not work either. 

I updated the original post to highlight the problem better, I did not pick an ideal title there.


## 4. Move workspace to monitor (via keybinding)
> 
> maybe creating 3 binds with swapactiveworkspaces? something like this:
> 
> ```
> bind = $mainMod TAB, 1, swapactiveworkspaces, 0 1
> bind = $mainMod TAB, 2, swapactiveworkspaces, 1 2
> bind = $mainMod TAB, 3, swapactiveworkspaces, 0 2
> ```
Not really feasible, I have ton of workspaces and there is not enough keys on the keyboard to do this :grin: That's why I wanted something  that allows me to work in a "relative" terms.

## 5. Waybar ignores changes of workspace names
> 
> don't know at which point are their releases, but using hyprland-git and waybar-git this works. there were changes about waybar modules and hyprland IPC lately...

Thanks, I will build from source & check :slightly_smiling_face:

**Comment by pbosab** [comment]
> Thanks for the reply 🙂
> ## 1. Multimonitor setup
> I know, I did that.
> 
> The thing is that my laptop is then configured with that "offset" of 6400 (as `monitor=eDP1,3840x2160@120,6400x0,1`) even if there are no other monitors attached.
> 
> And if I attach just a single monitor, it does not really make sense either - that's why I was wondering whether "dynamic" way is the correct approach here.

ok didn't understand that... yeah i think a "launcher" that change or load a different config will be needed then.

> ## 2. Workspace renaming / Cannot apply rules on renamed workspace
> I already have something that does the renaming, but thanks for this, I'll take a look!
> The problem is that _after renaming_ Hyprland stops applying rules from the config to that workspace 🙂 If I write `workspace = 1,....` it gets ignored, if I try `workspace = name:xxxxx` it does not work either.
> I updated the original post to highlight the problem better, I did not pick an ideal title there.

i think it's related to your question n°5

> ## 4. Move workspace to monitor (via keybinding)
> Not really feasible, I have ton of workspaces and there is not enough keys on the keyboard to do this 😁 That's why I wanted something that allows me to work in a "relative" terms.

those 0 1 and 2 are the monitors, not the workspaces:
the first bind swap the active workspaces between monitor 1 and 2
the second between monitor 2 and 3
and the third between monitor 1 and 3

> ## 5. Waybar ignores changes of workspace names
> Thanks, I will build from source & check 🙂

:+1:

**Reply by xoores** [reply]
## 1. Multimonitor setup
> > I know, I did that.
> > The thing is that my laptop is then configured with that "offset" of 6400 (as `monitor=eDP1,3840x2160@120,6400x0,1`) even if there are no other monitors attached.
> > And if I attach just a single monitor, it does not really make sense either - that's why I was wondering whether "dynamic" way is the correct approach here.

:+1: 

 ## 2. Workspace renaming / Cannot apply rules on renamed workspace
> > I already have something that does the renaming, but thanks for this, I'll take a look!
> > The problem is that _after renaming_ Hyprland stops applying rules from the config to that workspace 🙂 If I write `workspace = 1,....` it gets ignored, if I try `workspace = name:xxxxx` it does not work either.
> > I updated the original post to highlight the problem better, I did not pick an ideal title there.
> 
> i think it's related to your question n°5

Perhaps :shrug: I don't really know ATM.

## 4. Move workspace to monitor (via keybinding)
> > Not really feasible, I have ton of workspaces and there is not enough keys on the keyboard to do this 😁 That's why I wanted something that allows me to work in a "relative" terms.
> 
> those 0 1 and 2 are the monitors, not the workspaces: the first bind swap the active workspaces between monitor 1 and 2 the second between monitor 2 and 3 and the third between monitor 1 and 3

Ah. That seems a bit more clunky than simple "move to the left" and "move to the right" but it might be atleast a way to do that :sweat_smile:

**Comment by Aqa-Ib** [comment]
@xoores if you have a hybrid graphics with nvidia, then probably this is causing the delay when opening a program: https://github.com/hyprwm/Hyprland/issues/1065

**Reply by xoores** [reply]
Yeah, this might be it - that explains it :grin: I don't mind it *that much* if I know that this is a problem as I know that Hybrid Graphics is just a PITA in Linux unfortunately.

Thanks!

**Comment by xoores** [comment]
I just rebuilt Hyprland from the Master and I can confirm that points 2 & 5 are still present - will open an issue for those as it seems like a bug in Hyprland :slightly_smiling_face: 

edit: Opened as #3494

**Reply by pbosab** [reply]
i finally understood your problem at point n°2, tested it and isn't working for me too.
but for point n°5 i can confirm the name changes are working on waybar.
are you using the correct waybar module?
https://github.com/Alexays/Waybar/wiki/Module:-Hyprland#workspaces
did you set the correct format for the module? something like:
```
"hyprland/workspaces": {
	"format": "{name}",
}
```

**Reply by xoores** [reply]
Yup (atleast I think so):

```
    "hyprland/workspaces": {
        "all-outputs": false,
        "active-only": false,
        "on-click": "activate",
        "sort-by-name": false,
        "sort-by-number": true,
        "on-scroll-up": "hyprctl dispatch workspace e+1",
        "on-scroll-down": "hyprctl dispatch workspace e-1",
        "format": "{name}",
        "format-icons": {
            "default": ""
        },
    },
```

And when I do `hyprctl dispatch renameworkspace "1" ""` the Waybar still shows `1` as the workspace name.

**Reply by pbosab** [reply]
can you check 2 things?
do you have "hyprland-workspaces" in your waybar "modules-{left,right,center}" config?
can you try to edit the config to just leave:

```
"hyprland/workspaces": {
	"format": "{name}",
}
```

because i just disabled my renamer, used the `hyprctl dispatch renameworkspace "1" "test"` and worked...

**Reply by xoores** [reply]
I do: `"modules-left": ["hyprland/workspaces", "hyprland/submap"],`

I will have to check tomorrow as I have to do some work and I did not (yet) get my Hyprland to "fully working" mode :sweat_smile:  But thanks you for your time and help, I really appreciate it :)

**Reply by xoores** [reply]
I solved this by using AGS instead of waybar :slightly_smiling_face: It works over there and it is quite cool and way more capable!

Thanks for your help!

**Comment by xoores** [comment]
So I did not write for a while here and that is because unfortunately I had to revert back to X - Wayland has some weird problems when it comes to the reverse PRIME and I just could not work reliably when every plug/unplug of my docking station was a dice roll whether my computer will freeze :sweat_smile: 

So for now I will close this issue - thanks everyone for your help and input, I will try ocassionally and if/when the Wayland will be fixed, I will jump right back!

**Comment by stshow** [comment]
If I'm following the thread correctly, the idea is to mimic the i3wm behavior that does this:

```
# Move workspace to different monitor
bindsym $mod+ctrl+Shift+Up move workspace to output up
bindsym $mod+ctrl+Shift+Down move workspace to output down
bindsym $mod+ctrl+Shift+Left move workspace to output left
bindsym $mod+ctrl+Shift+Right move workspace to output right
```

I have successfully tested this in Hyprland with the following key binds in the `hyprland.conf` in `$HOME/.config/hypr`:

```
# Move workspace to monitor on left or right
bind = $mainMod Control, Left, movecurrentworkspacetomonitor, l
bind = $mainMod Control, Right, movecurrentworkspacetomonitor, r
```

This works for me. What I have not tested is the following:

```
# Move workspace to monitor on left or right
bind = $mainMod Control, Up, movecurrentworkspacetomonitor, u
bind = $mainMod Control, Down, movecurrentworkspacetomonitor, d
```

I hope that helps folks in the future.

**Reply by xoores** [reply]
It was quite some time ago and I had to switch back to i3/X so I don't really remember what I tried exactly 🤔 What I had was kind of hit and miss as it would for some reason get inverted (like I had some shortcuts working, then I would reboot my PC and instead of "move to the left screen" it would move my workspace to the right etc 😅

However I might have some time in coming weeks so I will try Hyprland once more as it was definitely the most cool WM I had so far 😎
